### PR TITLE
gh-94518: [_posixsubprocess] Replace variable validity flags with reserved values

### PR DESCRIPTION
**Pull request by arhadthedev**
Currently, `_posixsubprocess.c` uses group and user ids together with flags of whether the corrsponding variables were initialized. However such an implicit "either initialized or look somewhere else" confuses both a reader (another mental connection to constantly track between functions) and a compiler (warnings on potentially uninitialized variables). Instead, we can utilize a special group/user id as a flag value defined by POSIX but uses nowhere else. Namely:

- `gid`: `call_setgid = False` → `gid = -1`
- `uid`: `call_setuid = False` → `uid = -1`
- `groups`: `call_setgroups = False` → `groups = NULL` (obtained with `(groups_list != Py_None) ? groups : NULL`)

This PR is required for gh-94519.

---

PS.: @tiran I applied your suggestion from gh-94519 as this, separate PR to keep that one limited in scope and reviewable.

* Issue: gh-94518

### CODE REVIEW

**Comment by arhadthedev** [comment]
This change has no user-visible effect so no NEWS entry is required.

**Comment by arhadthedev** [comment]
> Please include a news entry.
>
> We typically only skip news if the change is a trivial internal change or change is already covered by a previous PR. Non-trivial commits get a news entry so other core devs, release manager, distributors, and users have an easier way to locate a problem when a change introduces a regression.

@tiran Done (initially I've thought this change is invisible to users so skipped it).

File: Modules/_posixsubprocess.c [hunk 1]
```diff
@@ -737,10 +737,7 @@ do_fork_exec(char *const exec_array[],
 
 #ifdef VFORK_USABLE
     if (child_sigmask) {
-        /* These are checked by our caller; verify them in debug builds. */
-        assert(!call_setuid);
-        assert(!call_setgid);
-        assert(!call_setgroups);
+        /* Checked by our caller; verify this in debug builds. */
```
Comment (gpshead) [inline Modules/_posixsubprocess.c]: keep equivalent assertions in here:

```
assert(uid == (uid_t)-1);
assert(gid == (gid_t)-1);
assert(groups_size < 0);
```

File: Modules/_posixsubprocess.c [hunk 2]
```diff
@@ -724,9 +724,9 @@ do_fork_exec(char *const exec_array[],
              int errpipe_read, int errpipe_write,
              int close_fds, int restore_signals,
              int call_setsid, pid_t pgid_to_set,
-             int call_setgid, gid_t gid,
-             int call_setgroups, size_t groups_size, const gid_t *groups,
-             int call_setuid, uid_t uid, int child_umask,
+             gid_t gid,
+             size_t groups_size, const gid_t *groups,
```
Comment (gpshead) [inline Modules/_posixsubprocess.c]: change groups_size to a Py_ssize_t as that is what we pass in. We can then use it being < 0 to mean the former `!call_setgroups`.  A call to C's `setgroups(0, NULL)` via an empty list of groups has meaning.

File: Modules/_posixsubprocess.c [hunk 3]
```diff
@@ -518,9 +518,9 @@ child_exec(char *const exec_array[],
            int errpipe_read, int errpipe_write,
            int close_fds, int restore_signals,
            int call_setsid, pid_t pgid_to_set,
-           int call_setgid, gid_t gid,
-           int call_setgroups, size_t groups_size, const gid_t *groups,
-           int call_setuid, uid_t uid, int child_umask,
+           gid_t gid,
+           size_t groups_size, const gid_t *groups,
```
Comment (gpshead) [inline Modules/_posixsubprocess.c]: Use Py_ssize_t for groups_size so that < 0 can mean "don't call setgroups".

File: Modules/_posixsubprocess.c [hunk 4]
```diff
@@ -619,17 +619,17 @@ child_exec(char *const exec_array[],
 #endif
 
 #ifdef HAVE_SETGROUPS
-    if (call_setgroups)
+    if (groups)
```
Comment (gpshead) [inline Modules/_posixsubprocess.c]: `if (groups_size >= 0)`

`setgroups(0, NULL)` is valid.

File: Modules/_posixsubprocess.c [hunk 5]
```diff
@@ -951,34 +946,31 @@ subprocess_fork_exec(PyObject *module, PyObject *args)
             }
```
Comment (gpshead) [inline Modules/_posixsubprocess.c:943]: Change

```
if ((groups = PyMem_RawMalloc(num_groups * sizeof(gid_t))) == NULL) {
```

to

```
if (num_groups > 0 && (groups = PyMem_RawMalloc(num_groups * sizeof(gid_t))) == NULL) {
```

File: Modules/_posixsubprocess.c [hunk 6]
```diff
@@ -1002,7 +994,7 @@ subprocess_fork_exec(PyObject *module, PyObject *args)
     /* Use vfork() only if it's safe. See the comment above child_exec(). */
     sigset_t old_sigs;
     if (preexec_fn == Py_None && allow_vfork &&
-        !call_setuid && !call_setgid && !call_setgroups) {
+        uid == (uid_t)-1 && gid == (gid_t)-1 && groups_list == Py_None) {
```
Comment (gpshead) [inline Modules/_posixsubprocess.c]: for consistency, use `num_groups < 0` here and below instead of the groups_list Py_None check.

**Review by gpshead — CHANGES_REQUESTED** [review]
I like this PR :)

**Comment by arhadthedev** [comment]
> `if (groups_size >= 0)`
>
> `setgroups(0, NULL) is valid.`

For some reason, `setgroups(0, NULL)` causes *Doctest*, *Check if generated files are up to date*, and SSL workflows to fail child processes with `PermissionError: [Errno 1] Operation not permitted` exception:

<details><summary>Stacktrace</summary>

```plain
Run ./python Lib/test/ssltests.py
Traceback (most recent call last):
OpenSSL 1.1.1n  15 Mar 2022
  File "/home/runner/work/cpython/cpython/Lib/test/ssltests.py", line 37, in <module>
    run_regrtests(*sys.argv[1:])
  File "/home/runner/work/cpython/cpython/Lib/test/ssltests.py", line 33, in run_regrtests
    result = subprocess.call(args)
             ^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/cpython/cpython/Lib/subprocess.py", line 378, in call
    with Popen(*popenargs, **kwargs) as p:
         ^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/home/runner/work/cpython/cpython/Lib/subprocess.py", line 1011, in __init__
    self._execute_child(args, executable, preexec_fn, close_fds,
  File "/home/runner/work/cpython/cpython/Lib/subprocess.py", line 1888, in _execute_child
    raise child_exception_type(errno_num, err_msg, err_filename)
PermissionError: [Errno 1] Operation not permitted
Error: Process completed with exit code 1.
```

</details>

However, this is fixed by excluding a zero from the comparison: `if (groups_size > 0)`. I think we may leave it as is because `setgroups(0, NULL)` is a no-op in `child_exec` (a positive returned value is dropped anyway):

> If size is zero, list is not modified, but the total number of supplementary group IDs for the process is returned.
>
> -- https://linux.die.net/man/2/setgroups

*For Bedevere: I have made the requested changes; please review again. For correct jump from Mentioned to this message instead of the following Bedevere's one: @gpshead.*

**Comment by arhadthedev** [comment]
@gpshead @tiran could you reconsider this PR please? This is a prerequirement for gh-94519, created to get rid of uninitialized variables.

**Comment by tiran** [comment]
We are currently busy with 3.11 release. It may take until 3.11.0 final before we come back to this patch. It's definitely on the list for 3.12.

**Comment by arhadthedev** [comment]
Todo: replace `groups_size`/`groups` parameter names with `extra_group_size`/`extra_groups` across the whole file to avoid confusion of passers-by. It's necessary to mitigate a misleading name of `setgroups` that supposes replacement of not supplimentary group affinities, but all ones. The mislead results in seeing the omission of `setgroups` call for `groups_size == 0` as a bug (like we should reset the inherited affinities but do nothing instead).

I'll do it in a separate PR because both this PR and gh-94519 are already big enough to clobber them with extra details.

**Comment by arhadthedev** [comment]
@tiran could you review and possibly merge this PR please? It's an implementation of your proposal https://github.com/python/cpython/pull/94519#issuecomment-1174734993 created as a separate PR to preserve diffs of both PRs readable.

**Review by gpshead — APPROVED** [review]
thanks for doing this!

**Comment by arhadthedev** [comment]
Thank you for merging, Gregory!

**Pull request merged**